### ISSUE TITLE
all_neighbors could be faster for SimpleDiGraph

### ISSUE THREAD

**Issue by abhinavmehndiratta** [JuliaGraphs/LightGraphs.jl#1295]
Here is the current definition for `all_neighbors` -
https://github.com/JuliaGraphs/LightGraphs.jl/blob/13ca5e013fd68082f84519844fef0e5b5a8803c6/src/core.jl#L282
SimpleGraphs store neighbors in sorted order, we could exploit that to make `union` faster, this would improve other algorithms. For eg. here is `kcore` with the new `all_neighbors`, I've just replaced `all_neighbors` with `new_all_neighbors` in the existing algorithm and you can see almost a 2X speedup - 

```julia
julia> using LightGraphs, BenchmarkTools

julia> g = SimpleDiGraph(15000, 90000)
{15000, 90000} directed simple Int64 graph

julia> function new_all_neighbors(g::SimpleDiGraph{T}, u::T) where T
           common = Vector{T}()
           i, j = 1, 1
           in_nbrs, out_nbrs = inneighbors(g, u), outneighbors(g, u)
           in_len, out_len = length(in_nbrs), length(out_nbrs)
           while i <= in_len && j <= out_len
               if in_nbrs[i] < out_nbrs[j]
                   push!(common, in_nbrs[i])
                   i += 1
               elseif in_nbrs[i] > out_nbrs[j]
                   push!(common, out_nbrs[j])
                   j += 1
               else
                   push!(common, in_nbrs[i])
                   i += 1
                   j += 1
               end
           end
           while i <= in_len
               push!(common, in_nbrs[i])
               i += 1    
           end
           while j <= out_len
               push!(common, out_nbrs[j])
               j += 1
           end
           return common
       end
new_all_neighbors (generic function with 1 method)

julia> function new_core_number(g::SimpleDiGraph{T}) where T
           has_self_loops(g) && throw(ArgumentError("graph must not have self-loops"))
           n = nv(g)    
           deg = T.(degree(g))
           maxdeg = maximum(deg)
           bin = zeros(T, maxdeg+1)
           vert = zeros(T, n)
           pos = zeros(T, n)
           for v = 1:n
               bin[deg[v]+1] += one(T)
           end
           start = one(T)
           for d = zero(T):maxdeg
               num = bin[d+1]
               bin[d+1] = start
               start += num
           end
           for v in vertices(g)
               pos[v] = bin[deg[v]+1]
               vert[pos[v]] = v
               bin[deg[v]+1] += one(T)
           end
           for d = maxdeg:-1:one(T)
              bin[d+1] = bin[d]
           end
           bin[1] = one(T)
           for i = 1:n
               v = vert[i]
               for u in new_all_neighbors(g, v) # CHANGE HERE
                   if deg[u] > deg[v]
                       du = deg[u]
                       pu = pos[u]
                       pw = bin[du+1]
                       w = vert[pw]
                       if u != w
                           pos[u] = pw
                           vert[pu] = w
                           pos[w] = pu
                           vert[pw] = u
                       end
                       bin[du+1] += one(T)
                       deg[u] -= one(T)
                   end
               end
           end
           return deg
       end
new_core_number (generic function with 1 method)

julia> core_number(g) == new_core_number(g)
true

julia> @benchmark core_number(g)
BenchmarkTools.Trial: 
  memory estimate:  20.94 MiB
  allocs estimate:  178446
  --------------
  minimum time:     29.600 ms (0.00% GC)
  median time:      33.099 ms (0.00% GC)
  mean time:        33.603 ms (5.09% GC)
  maximum time:     57.991 ms (8.49% GC)
  --------------
  samples:          149
  evals/sample:     1

julia> @benchmark new_core_number(g)
BenchmarkTools.Trial: 
  memory estimate:  5.55 MiB
  allocs estimate:  59088
  --------------
  minimum time:     13.191 ms (0.00% GC)
  median time:      13.855 ms (0.00% GC)
  mean time:        14.437 ms (3.02% GC)
  maximum time:     21.550 ms (0.00% GC)
  --------------
  samples:          346
  evals/sample:     1
```

**Comment by sbromberger** [comment]
Looks good:
```
julia> g = DiGraph(5_000_000, 20_000_000)
{5000000, 20000000} directed simple Int64 graph
julia> using BenchmarkTools
julia> i = rand(1:nv(g), 1_000_000);
julia> function f(g, i)
         @inbounds for v in i
           z = LightGraphs.SimpleGraphs.all_neighbors2(g, v)
         end
       end
f (generic function with 1 method)
julia> function h(g, i)
           @inbounds for v in i
             z = all_neighbors(g, v)
           end
         end
h (generic function with 1 method)

julia> @benchmark f($g, $i)
BenchmarkTools.Trial: 
  memory estimate:  247.74 MiB
  allocs estimate:  3311845
  --------------
  minimum time:     966.062 ms (0.00% GC)
  median time:      974.843 ms (0.00% GC)
  mean time:        1.003 s (3.10% GC)
  maximum time:     1.069 s (8.79% GC)
  --------------
  samples:          6
  evals/sample:     1

julia> @benchmark h($g, $i)
BenchmarkTools.Trial: 
  memory estimate:  887.96 MiB
  allocs estimate:  9865027
  --------------
  minimum time:     2.060 s (3.76% GC)
  median time:      2.066 s (3.78% GC)
  mean time:        2.066 s (3.79% GC)
  maximum time:     2.070 s (3.77% GC)
  --------------
  samples:          3
  evals/sample:     1
```

**Comment by sbromberger** [comment]
Could you make a PR to `src/SimpleGraphs/digraphs.jl` ?

**Comment by sbromberger** [comment]
I got it :)

Ref: #1298

**Comment by sbromberger** [comment]
Closed via #1298.
